src/scripting/lua_kernel_base.cpp

```diff
@@ -501,6 +511,179 @@ static int intf_get_language(lua_State* L)
+		lua_pushvalue(L, 1);
+		ON_SCOPE_EXIT(&) {
+			lua_settop(L, save_top);
+		};
+		do {
+			int table_idx = lua_absindex(L, -1);
+			for(lua_pushnil(L); lua_next(L, table_idx); lua_pop(L, 1)) {
+				if(lua_type(L, -2) == LUA_TSTRING) {
+					keys.push_back(lua_tostring(L,-2));
+				}
+			}
+			// Two possible exit cases:
+			// 1. getmetafield returns TNIL because there is no __index
+			// In this case, the stack is unchanged, so the while condition is still true.
+			// 2. The __index is not a table
+			// In this case, obviously the while condition fails
+			if(luaL_getmetafield(L, table_idx, "__index") == LUA_TNIL) break;
+		} while(lua_istable(L, -1));
+		if(lua_isfunction(L, -1)) {
+			lua_pop(L, 1);
+			dir_meta_helper(L, keys);
+		}
+	} else if(lua_isuserdata(L, 1) && !lua_islightuserdata(L, 1)) {
+		lua_pushvalue(L, 1);
+		dir_meta_helper(L, keys);
+		lua_pop(L, 1);
+	}
+	// Sort and remove any duplicates
+	std::sort(keys.begin(), keys.end());
+	auto final = std::unique(keys.begin(), keys.end());
+	final = std::remove_if(keys.begin(), final, [L](const std::string& key) {
+		if(key.compare(0, 2, "__") == 0) {
+			return true;
+		}
+		int save_top = lua_gettop(L);
+		ON_SCOPE_EXIT(&) {
+			lua_settop(L, save_top);
+		};
+		// Exclude deprecated elements
+		// Some keys may be write-only, which would raise an exception here
+		// In that case we just ignore it and assume not deprecated
+		// (the __dir metamethod would be responsible for excluding deprecated write-only keys)
+		lua_pushcfunction(L, impl_is_deprecated);
+		lua_pushvalue(L, 1);
+		lua_push(L, key);
+		if(lua_pcall(L, 2, 1, 0) == LUA_OK) {
+			return luaW_toboolean(L, -1);
+		}
+		return false;
+	});
+	keys.erase(final, keys.end());
+	size_t max_len = std::accumulate(keys.begin(), keys.end(), 0, [](size_t max, const std::string& next) {
+		return std::max(max, next.size());
+	});
+	// Let's limit to about 80 characters of total width with minimum 3 characters padding between columns
+	static const size_t MAX_WIDTH = 80, COL_PADDING = 3, SUFFIX_PADDING = 2;
+	size_t col_width = max_len + COL_PADDING + SUFFIX_PADDING;
+	size_t n_cols = (MAX_WIDTH + COL_PADDING) / col_width;
+	size_t n_rows = ceil(keys.size() / double(n_cols));
+	for(size_t i = 0; i < n_rows; i++) {
+		std::ostringstream line;
+		line.fill(' ');
+		line.setf(std::ios::left);
+		for(size_t j = 0; j < n_cols && j + (i * n_cols) < keys.size(); j++) {
+			int save_top = lua_gettop(L);
+			ON_SCOPE_EXIT(&) {
+				lua_settop(L, save_top);
+			};
+			lua_pushcfunction(L, impl_get_dir_suffix);
+			lua_pushvalue(L, 1);
+			const auto& key = keys[j + i * n_cols];
+			lua_pushlstring(L, key.c_str(), key.size());
+			std::string suffix = " !"; // Exclamation mark to indicate an error
+			if(lua_pcall(L, 2, 1, 0) == LUA_OK) {
+				suffix = luaL_checkstring(L, -1);
+			}
+			// This weird calculation is because width counts in bytes, not code points
+			// Since the suffix is a Unicode character, that messes up the alignment
```

Comment:
Can we be sure that there are no multibyte characters in the Lua names?

Reply:
That's a good point, as it's technically a possibility… however, Lua doesn't allow Unicode identifiers, and in 99% of cases these will be identifiers that are being listed. I'm not really sure if it's worth making the same correction for the actual identifiers…

Reply:
ok